lib/export-scala-regmap.js

```diff
@@ -0,0 +1,184 @@
+      case 'read-write':
+      case 'write-only':
+      case 'read-writeOnce':
+      case 'writeOnce':
+        direction = 'Output';
+        break;
+      case 'read-only':
+        direction = 'Input';
+        break;
+      default:
+        throw new Error(`invalid access field value: ${field.access}`);
+      }
+      return `val ${getScalaBundleRegFieldName(field, idx)} = ${direction}(UInt(${field.bits}.W))`;
+    } else {
+      return `// PAD ${field.bits}`;
+    }
+  });
+
+  return `class ${getScalaRegisterBundleName(register)} extends Bundle {
+${indent(2)(fields.join('\n'))}
+}`;
+};
+
+
+const getScalaAddressBlockBundleName = (addressBlock) => {
+  return `${addressBlock.name}AddressBlockBundle`;
+};
+
+const generateAddressBlockBundles = addressBlock => {
+  const registerBundles = addressBlock.registers.map((register) => {
+    return generateRegisterBundle(register);
+  });
+
+  const bundleFields = addressBlock.registers.map((register) => {
+    return `val ${register.name} = new ${getScalaRegisterBundleName(register)}()`;
+  });
+
+  return `${registerBundles.join('\n\n')}
+
+class ${getScalaAddressBlockBundleName(addressBlock)} extends Bundle {
+${indent(2)(bundleFields.join('\n'))}
+}`;
+};
+
+
+//// GNERATE REGROUTER ////
+
+const generateRegisterField = field => scalaRegisterName => {
+  if (field.access) {
+    const regFieldDesc = field.name || field.desc ?
+      `, RegFieldDesc(${field.name ? `"${field.name}"` : '""'}, ${field.desc ? `"${field.desc}"` : '""'}))` :
+      ')';
+
+    switch (field.access) {
+    case 'read-write':
+      return `RegField(${field.bits}, ${scalaRegisterName}${regFieldDesc}`;
+    case 'read-only':
+      return `RegField.r(${field.bits}, ${scalaRegisterName}${regFieldDesc}`;
+    case 'write-only':
+      return `RegField.w(${field.bits}, ${scalaRegisterName}${regFieldDesc}`;
+    case 'read-writeOnce':
+      return `RegField(${field.bits}, ${scalaRegisterName}, regFieldWriteOnceFn(${scalaRegisterName})${regFieldDesc}`;
+    case 'writeOnce':
+      return `RegField(${field.bits}, (), regFieldWriteOnceFn(${scalaRegisterName})${regFieldDesc}`;
+    default:
+      throw new Error(`invalid access field value: ${field.access}`);
+    }
+  } else {
+    return `RegField(${field.bits})`;
+  }
+};
+
+const generateRegister = register => scalaRegMapBundleName => {
+  const scalaRegBundleName = `${scalaRegMapBundleName}.${register.name}`;
+
+  const regFieldSeqElements = register.fields.map((field, idx) => {
+    const scalaBundleField = getScalaBundleRegFieldName(field, idx);
+    return generateRegisterField(field)(`${scalaRegBundleName}.${scalaBundleField}`);
+  });
+
+  const registerDesc = register.description ? `Some("${register.description}")` : 'None';
+
+  return `${register.addressOffset} -> RegFieldGroup("${register.name}", ${registerDesc}, Seq(
+${indent(2)(regFieldSeqElements.join(',\n'))}))`;
+};
+
+const generateBody = addressBlock => packageName => {
```

Comment:
Is there a reason you defined this as a curried function? At the only place where you actually call the function, you call it with both arguments simultaneously. I think it's much more typical in JavaScript to see multiple-parameter functions be defined as a single parameter list with multiple parameters rather than multiple, nested functions each taking a single parameter. Just to be absolutely clear, I am suggesting that we replace this with:

```suggestion
const generateBody = (addressBlock, packageName) => {
```

The same comment goes for most of your curried functions.

Reply:
I just like curried functions better. couldn't hurt right?

Reply:
we follow the same curry pattern for multiple functions with this scenario.
The reason is that we bind to the `addressBlock` first and then we can call in on multiple `packageName` in this address block.